app/src/main/java/ryey/easer/skills/operation/http_request/HttpRequestOperationData.java

```diff
@@ -0,0 +1,194 @@
+                    throw new IllegalStorageDataException(e);
+                }
+        }
+    }
+
+    @NonNull
+    @Override
+    public String serialize(@NonNull PluginDataFormat format) {
+        try {
+            JSONObject jsonObject = new JSONObject();
+            jsonObject.put(K_REQUEST_METHOD, requestMethod);
+            jsonObject.put(K_URL, url);
+            jsonObject.put(K_REQUEST_HEADER, requestHeader);
+            jsonObject.put(K_CONTENT_TYPE, contentType);
+            jsonObject.put(K_POST_DATA, postData);
+            return jsonObject.toString();
+        } catch (JSONException e) {
+            e.printStackTrace();
+            throw new IllegalStateException(e.getMessage());
+        }
+    }
+
+    @SuppressWarnings("RedundantIfStatement")
+    @Override
+    public boolean isValid() {
+        if (Utils.isBlank(url)) {
+            return false;
+        }
+
+        if (requestMethod == RequestMethod.POST && Utils.isBlank(contentType)) {
+            return false;
+        }
+
+        return true;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        HttpRequestOperationData that = (HttpRequestOperationData) o;
+        return requestMethod.equals(that.requestMethod) &&
+                url.equals(that.url) &&
+                requestHeader.equals(that.requestHeader) &&
+                contentType.equals(that.contentType) &&
+                postData.equals(that.postData);
+    }
+
+    @Override
+    public int hashCode() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
+            return Objects.hash(requestMethod, url, requestHeader, contentType, postData);
+        } else {
+            return requestHeader.hashCode() * 37 * 37 * 37 * 37
+                    + url.hashCode() * 37 * 37 * 37
+                    + requestHeader.hashCode() * 37 * 37
+                    + contentType.hashCode() * 37
+                    + postData.hashCode();
+        }
+    }
+
+    @Override
+    public int describeContents() {
+        return 0;
+    }
+
+    @Override
+    public void writeToParcel(Parcel dest, int flags) {
+        dest.writeSerializable(requestMethod);
+        dest.writeString(url);
+        dest.writeString(requestHeader);
+        dest.writeString(contentType);
+        dest.writeString(postData);
+    }
+
+    public static final Creator<HttpRequestOperationData> CREATOR = new Creator<HttpRequestOperationData>() {
+        public HttpRequestOperationData createFromParcel(Parcel in) {
+            return new HttpRequestOperationData(in);
+        }
+
+        public HttpRequestOperationData[] newArray(int size) {
+            return new HttpRequestOperationData[size];
+        }
+    };
+
+    private HttpRequestOperationData(Parcel in) {
+        requestMethod = (RequestMethod) Objects.requireNonNull(in.readSerializable());
```

Comment:
There is no need to do `Objects.requireNonNull()` when reading from the `Parcel`. Android will guarantee the correctness.

Reply:
The method is marked as @Nullable and it can indeed return null in [some cases](https://developer.android.com/reference/android/os/Parcel.html#readSerializable()).  I just added this because IDEA told me to, but I can also replace it with a skip-warning-annotation if that's better.

Reply:
From my understanding, it will return null iff the data does not exist in the `Parcel` -- which is impossible because you have already put that into the `Parcel` (in the `writeToParcel()` method).
Strange, my Android Studio never remind me of this.

Reply:
Do you use an older version of Android Studio (I have 3.5) or did you (accidentally) deactivated the inspection in the past?

![grafik](https://user-images.githubusercontent.com/577869/64489892-451d1680-d258-11e9-9acf-0012aab0334d.png)